keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,1060 @@
+bogged down.
+-->
+
+#### Story 1
+
+As a cluster admin, I want run some pods with privileged capabilities because
+the applications in the pods require it (e.g. `CAP_SYS_ADMIN` to mount a FUSE
+filesystem or `CAP_NET_ADMIN` to setup a VPN) but I don't want this extra
+capability to grant me any extra privilege on the host.
+
+#### Story 2
+
+As a cluster admin, I want to allow some pods to share the host user namespace
+if they need a feature only available in such user namespace, such as loading a kernel module with `CAP_SYS_MODULE`.
+
+### Notes/Constraints/Caveats
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Volumes Support
+
+The Linux kernel uses the effective user and group IDs (the ones the host) to
+check the file access permissions. Since with user namespaces IDs are mapped to
+a different value on the host, this causes issues accessing volumes if the
+pod is run with a different mapping, i.e. the effective user and group IDs on
+the host change.
+
+This proposal supports volume without changing the user and group IDs and leaves
+that problem to the user to manage. Future Linux kernel features such as shiftfs
+could allow different pods to see a volume with its own IDs but it is out of
+scope of this proposal. Among the possible future kernel solutions, we can list:
+
+- [shiftfs: uid/gid shifting filesystem](https://lwn.net/Articles/757650/)
+- [A new API for mounting filesystems](https://lwn.net/Articles/753473/)
+- [user_namespace: introduce fsid mappings](https://lwn.net/Articles/812221/)
+
+In regard to this proposal, volumes can be divided in ephemeral and non-ephemeral.
+
+Ephemeral volumes are associated to a **single** pod and their lifecyle is
+dependent on that pod. These are `configMap`, `secret`, `emptyDir`,
+`downwardAPI`, etc. These kind of volumes can work with any of the three
+different modes of `userNamespaceMode` as they are not shared by different pods
+and hence all the processes accessing those volumes have the same effective user
+and group IDs. Kubelet creates the files for those volumes and it can easily set
+the file ownership too.
+
+Non-ephemeral volumes are more difficult to support since they can be persistent
+and shared by multiple pods. This proposal supports volumes with two different
+strategies:
+- The `Cluster` mode makes it easier for pods to share files using volumes when those
+  don't have access permissions for `others` because the effective user and
+  group IDs on the host are the same for all the pods.
+- The semantics of semantics of `fsGroup` are respected, if it's specified it's
+  assumed to be the correct GID in the host and a 1-to-1 mapping entry for the
+  `fsGroup` is added to the GID mappings for the pod.
+
+This KEP doesn't impose any restriction on the different volumes and
```

Comment:
We should add examples on how users would achieve that with the current proposal.

Reply:
I reworded that section a bit on https://github.com/kinvolk/kubernetes-enhancements/pull/3/commits/78bfe606ec447025de706b8dc516c0cf0d675fd6.